plugin enhancement

zhouwei0115 opened the pull request:
file system service path enabled to save heap dump file;
error message given when required tools not present;
container path and local path to save dump file

TimGerlach commented on README.md:
```diff
@@ -87,13 +87,21 @@ OPTIONS:
 </pre>
 
-The heap dump or thread dump (depending on what you execute) will be outputted to `std-out`.
+the heap dump will be copied if -local-dir is specified(should be full folder path):
```
```suggestion
The heap dump will be copied if `-local-dir` is specified (should be full folder path):
```

TimGerlach commented on README.md:
````diff
@@ -87,13 +87,21 @@ OPTIONS:
+cf java heap-dump [my-app] --container-dir /var/fspath --local-dir /local/path
+```
+
+The thread dump will be outputted to `std-out`.
````
Do you mean that a thread dump will always be sent to std-out and the flag `--local-dir /local/path` is not supported for thread dumps?

zhouwei0115 replied:
no, thread-dump currently can't be saved to local file, as its output to stdout by jstack. commandExecutor didn't return it as output

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -168,10 +177,23 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	var remoteCommandTokens = []string{JavaDetectionCommand}
-
+	remoteFileFullPath := ""
+	localFileFullPath:= ""
```
It looks like the code doesn't wasn't formatted with a code formatter. Can you please apply [`gofmt`](https://pkg.go.dev/cmd/gofmt)?

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -168,10 +177,23 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	case heapDumpCommand:
-		heapdumpFileName := "/tmp/heapdump-" + uuidGenerator.Generate() + ".hprof"
+
+		valid, _ :=utils.CheckRequiredTools(applicationName)
```
Maybe `valid` isn't a very clear name. What about `supported`?

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,222 @@
+package utils
```
In think the `cfutils.go` should not output anything via `fmt.Fprintln` as it's a utility file that may be used in multiple places. If something fails, it should rather return an error object with an error message. With that, the caller can still decide whether to print the error message or to do something else instead.

zhouwei0115 replied:
updated accordingly, return the message as error.

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+func checkUserPathAvailability(app string, path string)(bool, error){
+	//fmt.Println("[[ -d \"" + path + "\" && -r \"" + path + "\" && -w \"" + path + "\" ]] && echo \"exists and read-writeable\"")
```
Please remove this line or use a logger with log level "DEBUG" instead.

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,222 @@
+	valid, _ := checkUserPathAvailability(data, userpath)
+
+	if valid {
+		//fmt.Println("userpath '" + userpath + "' will be used")
```
Please remove this line or use a logger with log level "DEBUG" instead.

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,222 @@
+		return "/tmp", nil
+	}
+
+	//fmt.Println(string(env[:]))
```
Please remove this line or use a logger with log level "DEBUG" instead.

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,222 @@
+	var cfAppEnv CFAppEnv
+	yaml.Unmarshal(env, &cfAppEnv)
+
+	//fmt.Println(result)
```
Please remove this line or use a logger with log level "DEBUG" instead.

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -189,26 +211,28 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+			//"fi",
 			"fi")
-		if !keepAfterDownload {
+		//if !keepAfterDownload {
```
Please remove comments

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -125,7 +127,9 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	commandFlags.NewBoolFlag("keep", "k", "whether to `keep` the heap/thread-dump on the container of the application instance after having downloaded it locally")
 	commandFlags.NewBoolFlag("dry-run", "n", "triggers the `dry-run` mode to show only the cf-ssh command that would have been executed")
-
+	commandFlags.NewStringFlag("container-dir", "cd", "specify folder path where the dump file should be stored in the container")
```
```suggestion
	commandFlags.NewStringFlag("container-dir", "cd", "specify the folder path where the dump file should be stored in the container")
```

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -125,7 +127,9 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	commandFlags.NewBoolFlag("dry-run", "n", "triggers the `dry-run` mode to show only the cf-ssh command that would have been executed")
-
+	commandFlags.NewStringFlag("container-dir", "cd", "specify folder path where the dump file should be stored in the container")
+	commandFlags.NewStringFlag("local-dir", "ld", "specify the folder where the dump file will be downloaded")
```
```suggestion
	commandFlags.NewStringFlag("local-dir", "ld", "specify the folder where the dump file will be downloaded to")
```

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -168,10 +177,23 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+		}
+		
+		fspath, _ := utils.GetAvailablePath(applicationName, remoteDir)
+		//fmt.Println(fspath)
```
Please remove this line or use a logger with log level "DEBUG" instead.

zhouwei0115 replied:
removed

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -189,26 +211,28 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
-			"cat "+heapdumpFileName,
-			"exit 0",
-			"fi",
+			//"echo "+heapdumpFileName,
```
Please remove comments

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -168,10 +177,23 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+		//fmt.Println(fspath)
+		dumpFile := applicationName + "-heapdump-" + uuidGenerator.Generate() + ".hprof"
+		heapdumpFileName := fspath + "/" + dumpFile
+		remoteFileFullPath = fspath + "/" + dumpFile
```
Why is `heapdumpFileName` and `remoteFileFullPath` the same?

zhouwei0115 replied:
removed duplicate

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -229,7 +253,26 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 
 	output, err := commandExecutor.Execute(cfSSHArguments)
+	
+	if strings.Compare(command, heapDumpCommand) == 0{
```
What about just `command == heapDumpCommand`?

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,222 @@
+}
+
+func readAppEnv(app string) ([]byte, error){
+	env, err := exec.Command("cf", "env", app).Output()
```
I think we should try to avoid parsing (potentially fragile) text output of the CF CLI.
The CF CLI supports multiple languages and we would need to deal with that, too.
Instead we should use `cf curl` instead and work with the returned JSON value.
You can find out which API calls are behind a command (e.g., cf env) by enabling tracing of the CF CLI and running the command again:

```shell
export CF_TRACE=true
cf env APP_NAME
```

For example, to get the actual JSON response from the cf env command, you can use:

```
cf curl /v3/apps/<APP_GUID>/env
```

zhouwei0115 replied:
Thanks for the comments and hint, helped me understanding it a lot.

TimGerlach reviewed (changes requested):
Thank you very much for this Pull-Request and the additional features it introduces.

I have made some comments. Please make sure to:
* use `gofmt` for code formatting
* remove all unnecessary comments
* remove `.DT_Store` from the commits (you can add it to the `.gitignore`
* add unit tests for `cfutils.go`
* check out other comments I made in the PR

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -229,7 +249,29 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	output, err := commandExecutor.Execute(cfSSHArguments)
+	if command == heapDumpCommand {
+
+		finalFile, err := util.FindDumpFile(applicationName, heapdumpFileName)
```
Why do we need to calculate the final heap dump file name again? Don't we have this already via `heapdumpFileName` which we constructed at line 200? Wouldn't it be sufficient if `FindDumpFile` solely checks whether the expected file has been generated or is there a situation where `heapdumpFileName != finalFile`?

zhouwei0115 replied:
seems that jvmmon doesn't support parameters to specify the path and name for the dump generated, so added this step in case the dump was generated by jvmmon.

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -125,6 +127,8 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	commandFlags.NewBoolFlag("keep", "k", "whether to `keep` the heap/thread-dump on the container of the application instance after having downloaded it locally")
 	commandFlags.NewBoolFlag("dry-run", "n", "triggers the `dry-run` mode to show only the cf-ssh command that would have been executed")
+	commandFlags.NewStringFlag("container-dir", "cd", "specify the folder path where the dump file should be stored in the container")
+	commandFlags.NewStringFlag("local-dir", "ld", "specify the folder where the dump file will be downloaded to")
```
We should explain that when `local-dir` is NOT given the heap dump will only be created in the container and not stored locally.

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -168,10 +182,24 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+		}
+		dumpFile := applicationName + "-heapdump-" + uuidGenerator.Generate() + ".hprof"
+		heapdumpFileName = fspath + "/" + dumpFile
+		localFileFullPath = localDir + "/" + dumpFile
```
As far as I can see we need `localFileFullPath` only in line 260 and only if the parameter `local-dir` was set. Maybe it's worth moving this variable closer to where it's being used (i.e. just before line 260)

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -229,7 +249,29 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+		if !keepAfterDownload {
+			err = util.DeleteRemoteFile(applicationName, heapdumpFileName)
+			if err != nil {
+				return "", err
```
Maybe it's worth adding:

```golang
fmt.Println("heap dump filed deleted in app container")
```

zhouwei0115 replied:
> I've added some more comments to go through.
> In general, we're missing to test the whole `utils/cfutils.go`. However, since I know that you're currently assigned with limited time, we can take this as a follow up improvement.
> 
> Before approving I'd like to quickly try out this build of the cf plugin myself with a demo Java app. I just saw that we have never added to the README how one can test the plugin locally. Can you maybe share how you've tested the plugin so that I do the test myself?

Summarized test cases from my side:
java-app: spring-music
file system mounted path: /var/fsjavadev

test case 1 - without any parameters set: 
command: cf java heap-dump spring-music
output: 
    heap dump will not be copied as parameter `local-dir` was not set
    heap dump filed deleted in app container
behind the ouput: monitor the container by ssh command 'cf ssh spring-music', will find during the plugin running, there will be one dump file with name 'spring-music-heapdump-<guid>.hprof' generated in /var/fsjavadev, and removed when the plugin exit running.

test case 2 - with parameter keep set:
command: cf java heap-dump spring-music -k
output:
    heap dump will not be copied as parameter `local-dir` was not set
behind the ouput: dump file should be kept in container under directory: /var/fsjavadev, we can check with command: 'cf ssh spring-music -c "ls -lrt /var/fsjavadev/*hprof"' and got result: 
-rw------- 1 vcap vcap 62270162 Aug 11 02:50 /var/fsjavadev/spring-music-heapdump-b9faef01-f061-4d9b-a0d0-785f741e4205.hprof

test case 3 - with invalid container dir set:
command: cf java heap-dump spring-music -container-dir /var/notexists
output:
    FAILED
    the container path specified doesn't exist or have no read and write access, please check and try again later

test case 4 - with valid container dir set:
command: cf java heap-dump spring-music -container-dir /var/fsjavadev
    heap dump will not be copied as parameter `local-dir` was not set
    heap dump filed deleted in app container
behind the ouput: monitor the container by ssh command 'cf ssh spring-music', will find during the plugin running, there will be one dump file with name 'spring-music-heapdump-<guid>.hprof' generated in /var/fsjavadev, and removed when the plugin exit running.

test case 5 - with valid container dir and keep set:
command: cf java heap-dump spring-music -container-dir /var/fsjavadev -keep
    heap dump will not be copied as parameter `local-dir` was not set
behind the ouput: dump file should be kept in container under directory: /var/fsjavadev, we can check with command: 'cf ssh spring-music -c "ls -lrt /var/fsjavadev/*hprof"' and got result: 
-rw------- 1 vcap vcap 62270162 Aug 11 02:50 /var/fsjavadev/spring-music-heapdump-b9faef01-f061-4d9b-a0d0-785f741e4205.hprof <--from test case 2
-rw------- 1 vcap vcap 62662145 Aug 11 02:59 /var/fsjavadev/spring-music-heapdump-9fd5d28c-3a60-4eda-8998-2f794d9fce29.hprof

test case 6 - with valid container dir and local-dir set:
command: cf java heap-dump spring-music -container-dir /tmp -local-dir /Users/i318048
    heap dump filed saved to: /Users/i318048/spring-music-heapdump-492b171a-5969-4b53-ae26-f4f0b2770c2c.hprof
    heap dump filed deleted in app container
behind the output: dump file was generated in /tmp and get deleted when plugin exits running

test case 7 - with valid container dir and invalid local-dir set:
command: cf java heap-dump spring-music -container-dir /tmp -local-dir /Users/invalid
output:
    FAILED
    Error creating local file at  /Users/invalid/spring-music-heapdump-91b34f7e-a3fd-42b0-af73-eb1f07f714e6.hprof. Please check that you are allowed to create files at the given local path.
behind the output: the dump file was generated under specified path /tmp, and user can copy the via a command like 'cf ssh spring-music -c "cat /tmp/spring-music-heapdump-<guid>.hprof" > /valid/local/path/file_name '


test case 8 - with local-dir set:
command: cf java heap-dump spring-music -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/spring-music-heapdump-860f7175-e42d-4f40-b7ee-9a651946735b.hprof
    heap dump filed deleted in app container

part 2, required tools not ready: 
java-app: firstappzw
file system: not present

test case 1: ssh disabled
command: cf java heap-dump firstappzw  -k 
output: 
    FAILED
    ssh is not enabled for app: 'firstappzw', please run below 2 shell commands to enable ssh and try again(please note application should be restarted before take effect):
    cf enable-ssh firstappzw
    cf restart firstappzw

test case 2: jvmmon and jmap missing(currently remove these two tools from container):
command: cf java heap-dump firstappzw  -k
output:
    FAILED
    jvmmon or jmap are required for generating heap dump, you can modify your application manifest.yaml on the 'JBP_CONFIG_OPEN_JDK_JRE' environment variable. This could be done like this:
                    ---
                    applications:
                    - name: <APP_NAME>
                    memory: 1G
                    path: <PATH_TO_BUILD_ARTIFACT>
                    buildpack: https://github.com/cloudfoundry/java-buildpack
                    env:
                            JBP_CONFIG_OPEN_JDK_JRE: '{ jre: { repository_root: "https://java-buildpack.cloudfoundry.org/openjdk-jdk/bionic/x86_64", version: 11.+ } }'

test case 3: with valid container dir and local dir set:
command: cf java heap-dump firstappzw -container-dir /tmp -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/firstappzw-heapdump-923e9722-423c-4724-702f-3e2b17132b5a.hprof
    heap dump filed deleted in app container
behind the output: dump file was generated in /tmp and get deleted when plugin exits running

test case 4: with valid local dir set:
command: cf java heap-dump firstappzw -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/firstappzw-heapdump-923e9722-423c-4724-702f-3e2b17132b5a.hprof
    heap dump filed deleted in app container
behind the output: without container directory specified, the dump file will be saved to file system attached, and to /tmp if file system not present

zhouwei0115 replied:
output message after deleted dump file.

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -229,7 +249,29 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+			} else {
+				return "", err
+			}
+		}
```
Maybe it's worth adding:
```golang
fmt.Println("heap dump will not be copied as parameter `local-dir` was not set")
```

zhouwei0115 replied:
updated

TimGerlach commented on cf_cli_java_plugin_test.go:
```diff
@@ -205,13 +208,73 @@ var _ = Describe("CfJavaPlugin", func() {
 
 			})
 
+			Context("with invalid container direcotry specified", func() {
```
```suggestion
			Context("with invalid container directory specified", func() {
```

TimGerlach commented on cf_cli_java_plugin_test.go:
```diff
@@ -205,13 +208,73 @@ var _ = Describe("CfJavaPlugin", func() {
+
+			})
+
+			Context("with invalid local direcotry specified", func() {
```
```suggestion
			Context("with invalid local directory specified", func() {
```

TimGerlach reviewed (changes requested):
I've added some more comments to go through.
In general, we're missing to test the whole `utils/cfutils.go`. However, since I know that you're currently assigned with limited time, we can take this as a follow up improvement.

Before approving I'd like to quickly try out this build of the cf plugin myself with a demo Java app. I just saw that we have never added to the README how one can test the plugin locally. Can you maybe share how you've tested the plugin so that I do the test myself?

zhouwei0115 commented:
> I've added some more comments to go through.
> In general, we're missing to test the whole `utils/cfutils.go`. However, since I know that you're currently assigned with limited time, we can take this as a follow up improvement.
> 
> Before approving I'd like to quickly try out this build of the cf plugin myself with a demo Java app. I just saw that we have never added to the README how one can test the plugin locally. Can you maybe share how you've tested the plugin so that I do the test myself?

##part 1, required tools present
java-app: spring-music
file system mounted path: /var/fsjavadev

#test case 1 - without any parameters set: 
command: cf java heap-dump spring-music
output: 
    heap dump will not be copied as parameter `local-dir` was not set
    heap dump filed deleted in app container
behind the ouput: monitor the container by ssh command 'cf ssh spring-music', will find during the plugin running, there will be one dump file with name 'spring-music-heapdump-<guid>.hprof' generated in /var/fsjavadev, and removed when the plugin exit running.

#test case 2 - with parameter keep set:
command: cf java heap-dump spring-music -k
output:
    heap dump will not be copied as parameter `local-dir` was not set
behind the ouput: dump file should be kept in container under directory: /var/fsjavadev, we can check with command: 'cf ssh spring-music -c "ls -lrt /var/fsjavadev/*hprof"' and got result: 
-rw------- 1 vcap vcap 62270162 Aug 11 02:50 /var/fsjavadev/spring-music-heapdump-b9faef01-f061-4d9b-a0d0-785f741e4205.hprof

#test case 3 - with invalid container dir set:
command: cf java heap-dump spring-music -container-dir /var/notexists
output:
    FAILED
    the container path specified doesn't exist or have no read and write access, please check and try again later

#test case 4 - with valid container dir set:
command: cf java heap-dump spring-music -container-dir /var/fsjavadev
    heap dump will not be copied as parameter `local-dir` was not set
    heap dump filed deleted in app container
behind the ouput: monitor the container by ssh command 'cf ssh spring-music', will find during the plugin running, there will be one dump file with name 'spring-music-heapdump-<guid>.hprof' generated in /var/fsjavadev, and removed when the plugin exit running.

#test case 5 - with valid container dir and keep set:
command: cf java heap-dump spring-music -container-dir /var/fsjavadev -keep
    heap dump will not be copied as parameter `local-dir` was not set
behind the ouput: dump file should be kept in container under directory: /var/fsjavadev, we can check with command: 'cf ssh spring-music -c "ls -lrt /var/fsjavadev/*hprof"' and got result: 
-rw------- 1 vcap vcap 62270162 Aug 11 02:50 /var/fsjavadev/spring-music-heapdump-b9faef01-f061-4d9b-a0d0-785f741e4205.hprof <--from test case 2
-rw------- 1 vcap vcap 62662145 Aug 11 02:59 /var/fsjavadev/spring-music-heapdump-9fd5d28c-3a60-4eda-8998-2f794d9fce29.hprof

#test case 6 - with valid container dir and local-dir set:
command: cf java heap-dump spring-music -container-dir /tmp -local-dir /Users/i318048
    heap dump filed saved to: /Users/i318048/spring-music-heapdump-492b171a-5969-4b53-ae26-f4f0b2770c2c.hprof
    heap dump filed deleted in app container
behind the output: dump file was generated in /tmp and get deleted when plugin exits running

#test case 7 - with valid container dir and invalid local-dir set:
command: cf java heap-dump spring-music -container-dir /tmp -local-dir /Users/invalid
output:
    FAILED
    Error creating local file at  /Users/invalid/spring-music-heapdump-91b34f7e-a3fd-42b0-af73-eb1f07f714e6.hprof. Please check that you are allowed to create files at the given local path.
behind the output: the dump file was generated under specified path /tmp, and user can copy the via a command like 'cf ssh spring-music -c "cat /tmp/spring-music-heapdump-<guid>.hprof" > /valid/local/path/file_name '


#test case 8 - with local-dir set:
command: cf java heap-dump spring-music -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/spring-music-heapdump-860f7175-e42d-4f40-b7ee-9a651946735b.hprof
    heap dump filed deleted in app container

##part 2, required tools not ready: 
java-app: firstappzw
file system: not present

#test case 1: ssh disabled
command: cf java heap-dump firstappzw  -k 
output: 
    FAILED
    ssh is not enabled for app: 'firstappzw', please run below 2 shell commands to enable ssh and try again(please note application should be restarted before take effect):
    cf enable-ssh firstappzw
    cf restart firstappzw

#test case 2: jvmmon and jmap missing(currently remove these two tools from container):
command: cf java heap-dump firstappzw  -k
output:
    FAILED
    jvmmon or jmap are required for generating heap dump, you can modify your application manifest.yaml on the 'JBP_CONFIG_OPEN_JDK_JRE' environment variable. This could be done like this:
                    ---
                    applications:
                    - name: <APP_NAME>
                    memory: 1G
                    path: <PATH_TO_BUILD_ARTIFACT>
                    buildpack: https://github.com/cloudfoundry/java-buildpack
                    env:
                            JBP_CONFIG_OPEN_JDK_JRE: '{ jre: { repository_root: "https://java-buildpack.cloudfoundry.org/openjdk-jdk/bionic/x86_64", version: 11.+ } }'

#test case 3: with valid container dir and local dir set:
command: cf java heap-dump firstappzw -container-dir /tmp -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/firstappzw-heapdump-923e9722-423c-4724-702f-3e2b17132b5a.hprof
    heap dump filed deleted in app container
behind the output: dump file was generated in /tmp and get deleted when plugin exits running

#test case 4: with valid local dir set:
command: cf java heap-dump firstappzw -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/firstappzw-heapdump-923e9722-423c-4724-702f-3e2b17132b5a.hprof
    heap dump filed deleted in app container
behind the output: without container directory specified, the dump file will be saved to file system attached, and to /tmp if file system not present

zhouwei0115 commented on utils/cfutils.go:
```diff
@@ -0,0 +1,204 @@
+func (checker CfJavaPluginUtilImpl) CopyOverCat(app string, src string, dest string) error {
+	f, err := os.OpenFile(dest, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+	if err != nil {
+		return errors.New("Error occured during create desination file: " + dest + ", please check you are allowed to create file in the path.")
```
updated as suggested

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,204 @@
+		return false, errors.New("ssh is not enabled for app: '" + app + "', please run below 2 shell commands to enable ssh and try again(please note application should be restarted before take effect):\ncf enable-ssh " + app + "\ncf restart " + app)
+	}
+
+	output, err = exec.Command("cf", "ssh", app, "-c", "find -executable -name jvmmon | head -1 | tr -d [:space:] | find -executable -name jmap | head -1 | tr -d [:space:]").Output()
```
This didn't work for my test app which uses the SAP Java Buildpack. In this command you require to have both `jvmmon` AND `jmap`. However, in apps deployed with the SAP Java Buildpack you only get `jvmmon`. In this case, the command will return an empty string.

A quick fix would be to check for either one:

```
find -executable -name jvmmon | head -1 | tr -d [:space:] || find -executable -name jmap | head -1 | tr -d [:space:]
```

zhouwei0115 replied:
updated the checking command.

zhouwei0115 commented on go.mod:
```diff
@@ -0,0 +1,29 @@
+module cf.plugin.ref/requires
```
its generated by go mod init

TimGerlach replied:
I see. According to the [go mod documentation](https://golang.org/doc/modules/gomod-ref#module-syntax), however, this should be set to the source code repository:

> The module path should be a path from which Go tools can download the module source. In practice, this is typically the module source's repository domain and path to the module code within the repository. The go command relies on this form when downloading module versions to resolve dependencies on the module user's behalf.
Even if you're not at first intending to make your module available for use from other code, using its repository path is a best practice that will help you avoid having to rename the module if you publish it later.

TimGerlach commented on utils/cfutils.go:
```diff
@@ -0,0 +1,203 @@
+		return false, errors.New("ssh is not enabled for app: '" + app + "', please run below 2 shell commands to enable ssh and try again(please note application should be restarted before take effect):\ncf enable-ssh " + app + "\ncf restart " + app)
+	}
+
+	output, err = exec.Command("cf", "ssh", app, "-c", "find -executable -name jvmmon | head -1 | tr -d [:space:] || find -executable -name jmap | head -1 | tr -d [:space:]").Output()
```
Unfortunately, this now works for the `sap_java_buildpack`, but not for the community Java buildpack anymore.
Although `jmap` exists:

```shell
$ find -executable -name jmap | head -1 | tr -d [:space:]
./app/.java-buildpack/open_jdk_jre/bin/jmap
```

it doesn't return any results if checked with the new command and hence exits:
```shell
$  find -executable -name jvmmon | head -1 | tr -d [:space:] || find -executable -name jmap | head -1 | tr -d [:space:]
[no result]
```

The problem is that `find` returns a zero exit code even if it doesn't find any file and therefore never executes the part after `||`.

Unfortunately, `find` doesn't support extended (`-E`) regex mode in the application container, so it's a little bit tricky to do it with plain `find`. What seems to work as a little workaround is using grep:

```shell
find -executable | grep -E '(.*jmap$)|(.*jvmmon$)'
```

TimGerlach commented on cf_cli_java_plugin.go:
```diff
@@ -229,7 +245,33 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
+		if err == nil && finalFile != "" {
+			heapdumpFileName = finalFile
+		}
+
```
When testing it, I was expecting to get some feedback from the plugin as to where it creates the heap dump in the container. This would help to find the heap dump in the container in case `-container-dir` is not provided. It also helps to understand whether the plugin properly detected the file system service path or not.

Maybe it's worth simply adding:
```golang
fmt.Println("successfully created heap dump in application container at: " + heapdumpFileName)
```

TimGerlach reviewed (changes requested):
I think we're almost done. Only few minor things left before we can merge it.
The plugin will significantly benefit from your valuable contribution, it's pretty cool to use it with the new options :)
Thank you for your ongoing support and patience with my slow reviews.

Please note that I will be on vacation until August 25 and will look at the PR again after my return.

zhouwei0115 commented:
Hi Tim,

When I tried modify it to like ‘module github.com/SAP/cf-cli-java-plugin’, compile will fail with error below:
go: updates to go.mod needed; to update it:
        go mod tidy
make: *** [compile] Error 1

Regards,
Wei


From: Tim Gerlach ***@***.***>
Reply-To: SAP/cf-cli-java-plugin ***@***.***>
Date: Friday, August 13, 2021 at 9:27 PM
To: SAP/cf-cli-java-plugin ***@***.***>
Cc: "Zhou, Wei" ***@***.***>, Author ***@***.***>
Subject: Re: [SAP/cf-cli-java-plugin] plugin enhancement (#24)


@TimGerlach commented on this pull request.

________________________________

In go.mod<https://github.com/SAP/cf-cli-java-plugin/pull/24#discussion_r688512959>:

> @@ -0,0 +1,29 @@

+module cf.plugin.ref/requires

I see. According to the go mod documentation<https://golang.org/doc/modules/gomod-ref#module-syntax>, however, this should be set to the source code repository:

The module path should be a path from which Go tools can download the module source. In practice, this is typically the module source's repository domain and path to the module code within the repository. The go command relies on this form when downloading module versions to resolve dependencies on the module user's behalf.
Even if you're not at first intending to make your module available for use from other code, using its repository path is a best practice that will help you avoid having to rename the module if you publish it later.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/SAP/cf-cli-java-plugin/pull/24#discussion_r688512959>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AVAOAUK5GJ3ELFCKDSOEBM3T4UMRZANCNFSM5BGOYJQA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.
